File: 473-matchsticks-to-square/473-matchsticks-to-square.py
class Solution:
    def makesquare(self, matchsticks: List[int]) -> bool:
        total = sum(matchsticks)
        
        target, rem = divmod(total, 4)
        
        if rem:
            return False
        
        matchsticks.sort(reverse=True)
        
        @cache
        def assign_sticks(index, top, bot, left, right) -> bool:
            nonlocal target
            
            sides = (top, bot, left, right)

            if all(side == target for side in sides):
                return True
            if any(side > target for side in sides):
                return False
            
            next_stick = matchsticks[index]
            
            top, bot, left, right = sorted(sides)
            
            return (
                assign_sticks(index + 1, top + next_stick, bot, left, right) or
                assign_sticks(index + 1, top, bot + next_stick, left, right) or
                assign_sticks(index + 1, top, bot, left + next_stick, right) or
                assign_sticks(index + 1, top, bot, left, right + next_stick)
            )
        
        return assign_sticks(0, 0, 0, 0, 0)
    

File: 473-matchsticks-to-square/473-matchsticks-to-square.go
func makesquare(matchsticks []int) bool {
    var total int
    
    for _, stick := range matchsticks {
        total += stick
    }
    
    if total % 4 != 0 {
        return false
    }
    
    target := total / 4
    
    sort.Slice(matchsticks, func(i, j int) bool {
        return matchsticks[i] > matchsticks[j]
    })
    
    var assignSticks func(index, top, bot, left, right int) bool
    
    assignSticks = func(index, top, bot, left, right int) bool {
        if top == target && bot == target && left == target && right == target {
            return true
        }
        if top > target || bot > target || left > target || right > target {
            return false
        }
        
        nextStick := matchsticks[index]
        return assignSticks(index + 1, top + nextStick, bot, left, right) ||
            assignSticks(index + 1, top, bot + nextStick, left, right) ||
            assignSticks(index + 1, top, bot, left + nextStick, right) ||
            assignSticks(index + 1, top, bot, left, right + nextStick)
    }
    
    return assignSticks(0, 0, 0, 0, 0)
}